modules/packages/OrderedSet/Treap.chpl

```diff
@@ -0,0 +1,862 @@
+      // Assuming direction is 1, we're finding the successor
+      if node == nil then return nil;
+      if node!.children[direction] {
+        // node has right child,
+        // find the leftmost node in its right child tree 
+        node = node!.children[direction];
+        direction ^= 1;
+        while node != nil && node!.children[direction] != nil {
+          node = node!.children[direction];
+        }
+        return node;
+      }
+      else {
+        // node doesn't have right child,
+        // find its first ancesstor whose left child tree it belongs to
+        direction ^= 1;
+        while node!.parent != nil && node!.parent!.children[direction] != node {
+          node = node!.parent;
+        }
+        return node!.parent;
+      }
+    }
+
+    pragma "no doc"
+    proc const _lower_bound(node: nodeType, e: eltType): nodeType {
+      if node == nil then return nil;
+      var cmp = _compare(e, node!.element);
+      if cmp == 0 then return node;
+      else if cmp < 0 {
+        var result = _lower_bound(node!.children[0], e);
+        if result != nil then return result;
+        else return node;
+      }
+      else return _lower_bound(node!.children[1], e);
+    }
+
+    pragma "no doc"
+    proc const _upper_bound(node: nodeType, e: eltType): nodeType {
+      if node == nil then return nil;
+      var cmp = _compare(e, node!.element);
+      if cmp >= 0 then return _upper_bound(node!.children[1], e);
+      else {
+        var result = _upper_bound(node!.children[0], e);
+        if result != nil then return result;
+        else return node;
+      }
+    }
+
+    /*
+      Find the first element in the orderedSet
+      which does not compare less than e.
+
+      Returns a tuple containing two element:
+      The first element is a `bool` that indicates whether there is such an element.
+      The second element is the occurence in the orderedSet, if there's any.
+
+      :returns: a tuple containing result
+      :rtype: `(bool, eltType)`
+    */
+    proc const lowerBound(e: eltType): (bool, eltType) {
+      if !isDefaultInitializable(e) {
+        compilerError("lowerBound is not avaliable on types that can't be \
+                      default-initialized, here: " + eltType: string);
+      }
+      
+      _enter(); defer _leave();
+
+      var node = _lower_bound(_root, e);
+      var result: (bool, eltType);
+      if node == nil then  result[0] = false;
+      else result = (true, node!.element);
+      return result;
+    }
+
+    /*
+      Find the first element in the orderedSet
+      which does not compare less than e.
+
+      Returns a tuple containing two element:
+      The first element is a `bool` that indicates whether there is such an element.
+      The second element is the occurence in the orderedSet, if there's any.
+
+      :returns: a tuple containing result
+      :rtype: `(bool, eltType)`
+    */
+    proc const upperBound(e: eltType): (bool, eltType) {
```

Comment:
Shouldn't the upper and lower bound return the ref.

Reply:
https://github.com/chapel-lang/chapel/issues/15921#issuecomment-663472904 Note that the way that it returns has changed here.